[jit] split canonicalize_ops, make a decompose pass

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #20039 dispatch and expose linear op
* **#19988 split canonicalize_ops, make a decompose pass**
* #19987 lower batchmm to non-diff optimization

Summary:
This PR enables the `addmm` decomposition happens later (after custom fusion, before batchmm), this will enable the custom fusion to see high level operator informations (such as `addmm` and `linear`, instead of `mm` + `add`)

Test plan:
Test if the decomposition pass works. 

Differential Revision: [D15190355](https://our.internmc.facebook.com/intern/diff/D15190355)

wanchaol committed:
split canonicalize_ops, make a decompose pass

wanchaol committed:
Update on "split canonicalize_ops, make a decompose pass"

split canonicalize_ops, make a decompose pass

gh-metadata: pytorch pytorch 19988 gh/wanchaol/2/head

wanchaol committed:
replace inlineGraph on "split canonicalize_ops, make a decompose pass"

split canonicalize_ops, make a decompose pass

gh-metadata: pytorch pytorch 19988 gh/wanchaol/2/head

bwasti commented on torch/csrc/jit/graph_executor.cpp:
```diff
@@ -645,6 +646,10 @@ struct GraphExecutorImpl {
+    DecomposeOps(graph);
+    // Autodiff and decompositon pass will replace some parts of graph with new graph
+    // these new graphs usually miss shape information on nodes, so we propagate shapes
+    // PropagateInputShapes(graph);
```
why is this commented out?

wanchaol replied:
some issues on `nn.Dropout` when enable this, I will need to look into that and will bring the pass back.

wanchaol committed:
reorg code, add linear AD formula on "[jit] split canonicalize_ops, make a decompose pass"

split canonicalize_ops, make a decompose pass

gh-metadata: pytorch pytorch 19988 gh/wanchaol/2/head

wanchaol committed:
Update on "[jit] split canonicalize_ops, make a decompose pass"

split canonicalize_ops, make a decompose pass

gh-metadata: pytorch pytorch 19988 gh/wanchaol/2/head

wanchaol committed:
Update on "[jit] split canonicalize_ops, make a decompose pass"

split canonicalize_ops, make a decompose pass

gh-metadata: pytorch pytorch 19988 gh/wanchaol/2/head

wanchaol committed:
update shape prop on "[jit] split canonicalize_ops, make a decompose pass"

split canonicalize_ops, make a decompose pass

gh-metadata: pytorch pytorch 19988 gh/wanchaol/2/head

wanchaol committed:
Update on "[jit] split canonicalize_ops, make a decompose pass"

split canonicalize_ops, make a decompose pass

gh-metadata: pytorch pytorch 19988 gh/wanchaol/2/head

zdevito commented on torch/csrc/jit/passes/decompose_ops.cpp:
```diff
@@ -0,0 +1,86 @@
+namespace torch {
+namespace jit {
+
+Value* decomposeOp(
```
What is this suppose to do? If it is suppose to execute this once, it doesn't work, it is allocating a std::once_flag each time through the function.

refactor:
```
static std::shared_ptr<Graph> addmm_graph = createAddMMGraph();
```

zdevito replied:
I would also prefer if we added a constructor to compilation unit so this can just be write. I don't want it to end up being the case that everyone copy/pastes this pattern.

static script::CompilationUnit addmm("....");

zdevito reviewed (commented):
I think the pass changes are good. Can you address the comments about using compilation unit to load code?

wanchaol committed:
update ad formula, add constructor to compilation unit on "[jit] split canonicalize_ops, make a decompose pass"

split canonicalize_ops, make a decompose pass

gh-metadata: pytorch pytorch 19988 gh/wanchaol/2/head

eellison commented on torch/csrc/jit/graph_executor.cpp:
```diff
@@ -638,11 +640,26 @@ struct GraphExecutorImpl {
+    // these new graphs usually consists of control flow and miss shape information on nodes, so we run shape
+    // prop and constant prop, as well as loop unrolling in some cases
+    PropagateInputShapes(graph);
+    ConstantPropagation(graph);
```
This should call peephole optimization as well i think.

wanchaol replied:
`runOptimization` have bunch of other passes that we don't need to run, just want to minimize the passes that needs to be run, plus it does not run `propagateInputShapes`, so I just make a separate group of passes.

eellison replied:
Could you call `propagateInputShapes` and then call into `runOptimization`? The downside is that we're now maintaining two separate sets of optimizations, and you might miss optimizations that you think you didn't need (as with prim::is_cuda because PeepholeOptimize wasn't called)

wanchaol replied:
sounds legit. the only concern i was having is that this will increase our compilation time quite a bit, that's why I want separate the passes in the first place. But given that we are potentially missing bunch of optimizations, I will change it to just run the optimization :)

wanchaol committed:
address comments on "[jit] split canonicalize_ops, make a decompose pass"

split canonicalize_ops, make a decompose pass

gh-metadata: pytorch pytorch 19988 gh/wanchaol/2/head

wanchaol committed:
fix tests on "[jit] split canonicalize_ops, make a decompose pass"

split canonicalize_ops, make a decompose pass

gh-metadata: pytorch pytorch 19988 gh/wanchaol/2/head

wanchaol committed:
change profiled shapes since we directly call the cpp function on "[jit] split canonicalize_ops, make a decompose pass"

split canonicalize_ops, make a decompose pass

gh-metadata: pytorch pytorch 19988 gh/wanchaol/2/head

wanchaol committed:
Update on "[jit] split canonicalize_ops, make a decompose pass"

split canonicalize_ops, make a decompose pass

gh-metadata: pytorch pytorch 19988 gh/wanchaol/2/head

wanchaol committed:
Update on "[jit] split canonicalize_ops, make a decompose pass"

split canonicalize_ops, make a decompose pass

gh-metadata: pytorch pytorch 19988 gh/wanchaol/2/head
